File: datastructures/linkedlist/python/algorithms/classic/reverse-linked-list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def print_list(head):
    current = head

    while current:
        print(current.val)
        current = current.next


def reverse_list(head):
    prev = None

    while head:
        next = head.next
        head.next = prev
        prev = head
        head = next

    return prev


if __name__ == "__main__":
    node_1 = ListNode(1)
    node_2 = ListNode(2)
    node_3 = ListNode(3)
    node_4 = ListNode(4)

    node_1.next = node_2
    node_2.next = node_3
    node_3.next = node_4

    print_list(node_1)

    reversed = reverse_list(node_1)

    print("Reversed Linked List")
    print_list(reversed)

File: datastructures/linkedlist/java/algorithms/classic/ReverseLinkedList.java
public class ReverseLinkedList {

    public static ListNode2 reverseList(ListNode2 head) {
        ListNode2 prev = null;

        while (head != null) {
            ListNode2 next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }

        return prev;
    }

    private static void print(ListNode2 head) {
        ListNode2 current = head;

        while (current != null) {
            System.out.print(current.val + " ");
            current = current.next;
        }
    }

    public static void main(String[] args) {
        ListNode2 node4 = new ListNode2(4, null);
        ListNode2 node3 = new ListNode2(3, node4);
        ListNode2 node2 = new ListNode2(2, node3);
        ListNode2 node1 = new ListNode2(1, node2);

        print(node1);

        ListNode2 reversed = reverseList(node1);

        System.out.println("\n Reversed Linked List: ");
        print(reversed);
    }
}
